Kernel/Heap/Heap.h

```diff
@@ -68,40 +68,56 @@ class Heap {
         return needed_chunks * CHUNK_SIZE + (needed_chunks + 7) / 8;
     }
 
-    void* allocate(size_t size, CallerWillInitializeMemory caller_will_initialize_memory)
+    void* allocate(size_t size, size_t alignment, CallerWillInitializeMemory caller_will_initialize_memory)
     {
+        // We can't align anything less than CHUNK_SIZE, presumably due to edge cases in the bitmap calculation.
+        // Not that it really matters, we track only chunks, nothing smaller.
+        if (alignment < CHUNK_SIZE)
+            alignment = CHUNK_SIZE;
+
         // We need space for the AllocationHeader at the head of the block.
         size_t real_size = size + sizeof(AllocationHeader);
         size_t chunks_needed = (real_size + CHUNK_SIZE - 1) / CHUNK_SIZE;
+        size_t chunk_alignment = (alignment + CHUNK_SIZE - 1) / CHUNK_SIZE;
 
         if (chunks_needed > free_chunks())
             return nullptr;
 
         Optional<size_t> first_chunk;
 
         // Choose the right policy for allocation.
+        // FIXME: These should utilize the alignment directly instead of trying to allocate `size + alignment`.
         constexpr u32 best_fit_threshold = 128;
         if (chunks_needed < best_fit_threshold) {
-            first_chunk = m_bitmap.find_first_fit(chunks_needed);
+            first_chunk = m_bitmap.find_first_fit(chunks_needed + chunk_alignment);
         } else {
-            first_chunk = m_bitmap.find_best_fit(chunks_needed);
+            first_chunk = m_bitmap.find_best_fit(chunks_needed + chunk_alignment);
         }
 
         if (!first_chunk.has_value())
             return nullptr;
 
         auto* a = (AllocationHeader*)(m_chunks + (first_chunk.value() * CHUNK_SIZE));
+
+        // Align the starting address and verify that we haven't gone outside the calculated free area.
+        a = (AllocationHeader*)((FlatPtr)a + alignment - (FlatPtr)a->data % alignment);
+        auto aligned_first_chunk = ((FlatPtr)a - (FlatPtr)m_chunks) / CHUNK_SIZE;
+        VERIFY(first_chunk.value() <= aligned_first_chunk && aligned_first_chunk <= first_chunk.value() + chunk_alignment);
```

Comment:
In general, I think it's nicer to break up `VERIFY(a && b)` into `VERIFY(a); VERIFY(b);`

Reply:
I removed the entire second condition, as we technically only care that we don't start before the free area starts (first condition) and that we don't end after the end of the free area (third condition). The second condition was just the same as the third condition with `+ chunks_needed` removed on either side.